
<!doctype html>
<html lang="en" class="no-js">
  <head>
    
      <meta charset="utf-8">
      <meta name="viewport" content="width=device-width,initial-scale=1">
      
      
      
      
        <link rel="prev" href="../global-config/">
      
      
        <link rel="next" href="../input-formatting/">
      
      
      <link rel="icon" href="../assets/images/favicon.png">
      <meta name="generator" content="mkdocs-1.6.1, mkdocs-material-9.6.7">
    
    
      
        <title>Starting an analysis - ViroConstrictor</title>
      
    
    
      <link rel="stylesheet" href="../assets/stylesheets/main.8608ea7d.min.css">
      
        
        <link rel="stylesheet" href="../assets/stylesheets/palette.06af60db.min.css">
      
      


    
    
      
    
    
      
        
        
        <link rel="preconnect" href="https://fonts.gstatic.com" crossorigin>
        <link rel="stylesheet" href="https://fonts.googleapis.com/css?family=Roboto:300,300i,400,400i,700,700i%7CRoboto+Mono:400,400i,700,700i&display=fallback">
        <style>:root{--md-text-font:"Roboto";--md-code-font:"Roboto Mono"}</style>
      
    
    
    <script>__md_scope=new URL("..",location),__md_hash=e=>[...e].reduce(((e,_)=>(e<<5)-e+_.charCodeAt(0)),0),__md_get=(e,_=localStorage,t=__md_scope)=>JSON.parse(_.getItem(t.pathname+"."+e)),__md_set=(e,_,t=localStorage,a=__md_scope)=>{try{t.setItem(a.pathname+"."+e,JSON.stringify(_))}catch(e){}}</script>
    
      

    
    
    
  </head>
  
  
    
    
      
    
    
    
    
    <body dir="ltr" data-md-color-scheme="default" data-md-color-primary="indigo" data-md-color-accent="light-blue">
  
    
    <input class="md-toggle" data-md-toggle="drawer" type="checkbox" id="__drawer" autocomplete="off">
    <input class="md-toggle" data-md-toggle="search" type="checkbox" id="__search" autocomplete="off">
    <label class="md-overlay" for="__drawer"></label>
    <div data-md-component="skip">
      
        
        <a href="#starting-an-analysis" class="md-skip">
          Skip to content
        </a>
      
    </div>
    <div data-md-component="announce">
      
    </div>
    
      <div data-md-color-scheme="default" data-md-component="outdated" hidden>
        
      </div>
    
    
      

  

<header class="md-header md-header--shadow md-header--lifted" data-md-component="header">
  <nav class="md-header__inner md-grid" aria-label="Header">
    <a href=".." title="ViroConstrictor" class="md-header__button md-logo" aria-label="ViroConstrictor" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M4 2h2v2c0 1.44.68 2.61 1.88 3.78.86.83 2.01 1.63 3.21 2.42l-1.83 1.19C8.27 10.72 7.31 10 6.5 9.21 5.07 7.82 4 6.1 4 4zm14 0h2v2c0 2.1-1.07 3.82-2.5 5.21-1.41 1.38-3.21 2.52-4.96 3.63-1.75 1.12-3.45 2.21-4.66 3.38C6.68 17.39 6 18.56 6 20v2H4v-2c0-2.1 1.07-3.82 2.5-5.21 1.41-1.38 3.21-2.52 4.96-3.63 1.75-1.12 3.45-2.21 4.66-3.38C17.32 6.61 18 5.44 18 4zm-3.26 10.61c.99.67 1.95 1.39 2.76 2.18C18.93 16.18 20 17.9 20 20v2h-2v-2c0-1.44-.68-2.61-1.88-3.78-.86-.83-2.01-1.63-3.21-2.42zM7 3h10v1l-.06.5H7.06L7 4zm.68 3h8.64c-.24.34-.52.69-.9 1.06l-.51.44H9.07l-.49-.44c-.38-.37-.66-.72-.9-1.06m1.41 10.5h5.84l.49.44c.38.37.66.72.9 1.06H7.68c.24-.34.52-.69.9-1.06zm-2.03 3h9.88l.06.5v1H7v-1z"/></svg>

    </a>
    <label class="md-header__button md-icon" for="__drawer">
      
      <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M3 6h18v2H3zm0 5h18v2H3zm0 5h18v2H3z"/></svg>
    </label>
    <div class="md-header__title" data-md-component="header-title">
      <div class="md-header__ellipsis">
        <div class="md-header__topic">
          <span class="md-ellipsis">
            ViroConstrictor
          </span>
        </div>
        <div class="md-header__topic" data-md-component="header-topic">
          <span class="md-ellipsis">
            
              Starting an analysis
            
          </span>
        </div>
      </div>
    </div>
    
      
        <form class="md-header__option" data-md-component="palette">
  
    
    
    
    <input class="md-option" data-md-color-media="(prefers-color-scheme: light)" data-md-color-scheme="default" data-md-color-primary="indigo" data-md-color-accent="light-blue"  aria-label="Switch to dark mode"  type="radio" name="__palette" id="__palette_0">
    
      <label class="md-header__button md-icon" title="Switch to dark mode" for="__palette_1" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M17 6H7c-3.31 0-6 2.69-6 6s2.69 6 6 6h10c3.31 0 6-2.69 6-6s-2.69-6-6-6m0 10H7c-2.21 0-4-1.79-4-4s1.79-4 4-4h10c2.21 0 4 1.79 4 4s-1.79 4-4 4M7 9c-1.66 0-3 1.34-3 3s1.34 3 3 3 3-1.34 3-3-1.34-3-3-3"/></svg>
      </label>
    
  
    
    
    
    <input class="md-option" data-md-color-media="(prefers-color-scheme: dark)" data-md-color-scheme="slate" data-md-color-primary="red" data-md-color-accent="orange"  aria-label="Switch light mode"  type="radio" name="__palette" id="__palette_1">
    
      <label class="md-header__button md-icon" title="Switch light mode" for="__palette_0" hidden>
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M17 7H7a5 5 0 0 0-5 5 5 5 0 0 0 5 5h10a5 5 0 0 0 5-5 5 5 0 0 0-5-5m0 8a3 3 0 0 1-3-3 3 3 0 0 1 3-3 3 3 0 0 1 3 3 3 3 0 0 1-3 3"/></svg>
      </label>
    
  
</form>
      
    
    
      <script>var palette=__md_get("__palette");if(palette&&palette.color){if("(prefers-color-scheme)"===palette.color.media){var media=matchMedia("(prefers-color-scheme: light)"),input=document.querySelector(media.matches?"[data-md-color-media='(prefers-color-scheme: light)']":"[data-md-color-media='(prefers-color-scheme: dark)']");palette.color.media=input.getAttribute("data-md-color-media"),palette.color.scheme=input.getAttribute("data-md-color-scheme"),palette.color.primary=input.getAttribute("data-md-color-primary"),palette.color.accent=input.getAttribute("data-md-color-accent")}for(var[key,value]of Object.entries(palette.color))document.body.setAttribute("data-md-color-"+key,value)}</script>
    
    
    
      <label class="md-header__button md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
      </label>
      <div class="md-search" data-md-component="search" role="dialog">
  <label class="md-search__overlay" for="__search"></label>
  <div class="md-search__inner" role="search">
    <form class="md-search__form" name="search">
      <input type="text" class="md-search__input" name="query" aria-label="Search" placeholder="Search" autocapitalize="off" autocorrect="off" autocomplete="off" spellcheck="false" data-md-component="search-query" required>
      <label class="md-search__icon md-icon" for="__search">
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M9.5 3A6.5 6.5 0 0 1 16 9.5c0 1.61-.59 3.09-1.56 4.23l.27.27h.79l5 5-1.5 1.5-5-5v-.79l-.27-.27A6.52 6.52 0 0 1 9.5 16 6.5 6.5 0 0 1 3 9.5 6.5 6.5 0 0 1 9.5 3m0 2C7 5 5 7 5 9.5S7 14 9.5 14 14 12 14 9.5 12 5 9.5 5"/></svg>
        
        <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M20 11v2H8l5.5 5.5-1.42 1.42L4.16 12l7.92-7.92L13.5 5.5 8 11z"/></svg>
      </label>
      <nav class="md-search__options" aria-label="Search">
        
        <button type="reset" class="md-search__icon md-icon" title="Clear" aria-label="Clear" tabindex="-1">
          
          <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M19 6.41 17.59 5 12 10.59 6.41 5 5 6.41 10.59 12 5 17.59 6.41 19 12 13.41 17.59 19 19 17.59 13.41 12z"/></svg>
        </button>
      </nav>
      
    </form>
    <div class="md-search__output">
      <div class="md-search__scrollwrap" tabindex="0" data-md-scrollfix>
        <div class="md-search-result" data-md-component="search-result">
          <div class="md-search-result__meta">
            Initializing search
          </div>
          <ol class="md-search-result__list" role="presentation"></ol>
        </div>
      </div>
    </div>
  </div>
</div>
    
    
      <div class="md-header__source">
        <a href="https://github.com/RIVM-bioinformatics/ViroConstrictor" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.7.2 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81"/></svg>
  </div>
  <div class="md-source__repository">
    RIVM-bioinformatics/ViroConstrictor
  </div>
</a>
      </div>
    
  </nav>
  
    
      
<nav class="md-tabs" aria-label="Tabs" data-md-component="tabs">
  <div class="md-grid">
    <ul class="md-tabs__list">
      
        
  
  
  
    <li class="md-tabs__item">
      <a href=".." class="md-tabs__link">
        
  
    
  
  Home

      </a>
    </li>
  

      
        
  
  
  
    
    
      <li class="md-tabs__item">
        <a href="../installation/" class="md-tabs__link">
          
  
  Getting started

        </a>
      </li>
    
  

      
        
  
  
    
  
  
    
    
      <li class="md-tabs__item md-tabs__item--active">
        <a href="./" class="md-tabs__link">
          
  
  Usage

        </a>
      </li>
    
  

      
        
  
  
  
    <li class="md-tabs__item">
      <a href="../changelog/" class="md-tabs__link">
        
  
    
  
  Changelog

      </a>
    </li>
  

      
    </ul>
  </div>
</nav>
    
  
</header>
    
    <div class="md-container" data-md-component="container">
      
      
        
      
      <main class="md-main" data-md-component="main">
        <div class="md-main__inner md-grid">
          
            
              
              <div class="md-sidebar md-sidebar--primary" data-md-component="sidebar" data-md-type="navigation" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    


  


<nav class="md-nav md-nav--primary md-nav--lifted" aria-label="Navigation" data-md-level="0">
  <label class="md-nav__title" for="__drawer">
    <a href=".." title="ViroConstrictor" class="md-nav__button md-logo" aria-label="ViroConstrictor" data-md-component="logo">
      
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M4 2h2v2c0 1.44.68 2.61 1.88 3.78.86.83 2.01 1.63 3.21 2.42l-1.83 1.19C8.27 10.72 7.31 10 6.5 9.21 5.07 7.82 4 6.1 4 4zm14 0h2v2c0 2.1-1.07 3.82-2.5 5.21-1.41 1.38-3.21 2.52-4.96 3.63-1.75 1.12-3.45 2.21-4.66 3.38C6.68 17.39 6 18.56 6 20v2H4v-2c0-2.1 1.07-3.82 2.5-5.21 1.41-1.38 3.21-2.52 4.96-3.63 1.75-1.12 3.45-2.21 4.66-3.38C17.32 6.61 18 5.44 18 4zm-3.26 10.61c.99.67 1.95 1.39 2.76 2.18C18.93 16.18 20 17.9 20 20v2h-2v-2c0-1.44-.68-2.61-1.88-3.78-.86-.83-2.01-1.63-3.21-2.42zM7 3h10v1l-.06.5H7.06L7 4zm.68 3h8.64c-.24.34-.52.69-.9 1.06l-.51.44H9.07l-.49-.44c-.38-.37-.66-.72-.9-1.06m1.41 10.5h5.84l.49.44c.38.37.66.72.9 1.06H7.68c.24-.34.52-.69.9-1.06zm-2.03 3h9.88l.06.5v1H7v-1z"/></svg>

    </a>
    ViroConstrictor
  </label>
  
    <div class="md-nav__source">
      <a href="https://github.com/RIVM-bioinformatics/ViroConstrictor" title="Go to repository" class="md-source" data-md-component="source">
  <div class="md-source__icon md-icon">
    
    <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.7.2 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2024 Fonticons, Inc.--><path d="M439.55 236.05 244 40.45a28.87 28.87 0 0 0-40.81 0l-40.66 40.63 51.52 51.52c27.06-9.14 52.68 16.77 43.39 43.68l49.66 49.66c34.23-11.8 61.18 31 35.47 56.69-26.49 26.49-70.21-2.87-56-37.34L240.22 199v121.85c25.3 12.54 22.26 41.85 9.08 55a34.34 34.34 0 0 1-48.55 0c-17.57-17.6-11.07-46.91 11.25-56v-123c-20.8-8.51-24.6-30.74-18.64-45L142.57 101 8.45 235.14a28.86 28.86 0 0 0 0 40.81l195.61 195.6a28.86 28.86 0 0 0 40.8 0l194.69-194.69a28.86 28.86 0 0 0 0-40.81"/></svg>
  </div>
  <div class="md-source__repository">
    RIVM-bioinformatics/ViroConstrictor
  </div>
</a>
    </div>
  
  <ul class="md-nav__list" data-md-scrollfix>
    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href=".." class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Home
    
  </span>
  

      </a>
    </li>
  

    
      
      
  
  
  
  
    
    
    
      
      
    
    
    <li class="md-nav__item md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_2" >
        
          
          <label class="md-nav__link" for="__nav_2" id="__nav_2_label" tabindex="0">
            
  
  <span class="md-ellipsis">
    Getting started
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_2_label" aria-expanded="false">
          <label class="md-nav__title" for="__nav_2">
            <span class="md-nav__icon md-icon"></span>
            Getting started
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../installation/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Download & installation
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../global-config/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Global configuration
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
    
  
  
  
    
    
    
      
        
        
      
      
    
    
    <li class="md-nav__item md-nav__item--active md-nav__item--section md-nav__item--nested">
      
        
        
        <input class="md-nav__toggle md-toggle " type="checkbox" id="__nav_3" checked>
        
          
          <label class="md-nav__link" for="__nav_3" id="__nav_3_label" tabindex="">
            
  
  <span class="md-ellipsis">
    Usage
    
  </span>
  

            <span class="md-nav__icon md-icon"></span>
          </label>
        
        <nav class="md-nav" data-md-level="1" aria-labelledby="__nav_3_label" aria-expanded="true">
          <label class="md-nav__title" for="__nav_3">
            <span class="md-nav__icon md-icon"></span>
            Usage
          </label>
          <ul class="md-nav__list" data-md-scrollfix>
            
              
                
  
  
    
  
  
  
    <li class="md-nav__item md-nav__item--active">
      
      <input class="md-nav__toggle md-toggle" type="checkbox" id="__toc">
      
      
        
      
      
        <label class="md-nav__link md-nav__link--active" for="__toc">
          
  
  <span class="md-ellipsis">
    Starting an analysis
    
  </span>
  

          <span class="md-nav__icon md-icon"></span>
        </label>
      
      <a href="./" class="md-nav__link md-nav__link--active">
        
  
  <span class="md-ellipsis">
    Starting an analysis
    
  </span>
  

      </a>
      
        

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#overview-of-command-line-options" class="md-nav__link">
    <span class="md-ellipsis">
      Overview of command-line options
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#run-a-single-target-analysis" class="md-nav__link">
    <span class="md-ellipsis">
      Run a single-target analysis
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#run-a-multi-target-analysis" class="md-nav__link">
    <span class="md-ellipsis">
      Run a multi-target analysis
    </span>
  </a>
  
</li>
      
    </ul>
  
</nav>
      
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../input-formatting/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Formatting inputs
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../presets/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Working with analysis presets
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../amplicons/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Explanation of amplicon types
    
  </span>
  

      </a>
    </li>
  

              
            
              
                
  
  
  
  
    <li class="md-nav__item">
      <a href="../multi-reference/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Working with multiple references
    
  </span>
  

      </a>
    </li>
  

              
            
          </ul>
        </nav>
      
    </li>
  

    
      
      
  
  
  
  
    <li class="md-nav__item">
      <a href="../changelog/" class="md-nav__link">
        
  
  <span class="md-ellipsis">
    Changelog
    
  </span>
  

      </a>
    </li>
  

    
  </ul>
</nav>
                  </div>
                </div>
              </div>
            
            
              
              <div class="md-sidebar md-sidebar--secondary" data-md-component="sidebar" data-md-type="toc" >
                <div class="md-sidebar__scrollwrap">
                  <div class="md-sidebar__inner">
                    

<nav class="md-nav md-nav--secondary" aria-label="Table of contents">
  
  
  
    
  
  
    <label class="md-nav__title" for="__toc">
      <span class="md-nav__icon md-icon"></span>
      Table of contents
    </label>
    <ul class="md-nav__list" data-md-component="toc" data-md-scrollfix>
      
        <li class="md-nav__item">
  <a href="#overview-of-command-line-options" class="md-nav__link">
    <span class="md-ellipsis">
      Overview of command-line options
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#run-a-single-target-analysis" class="md-nav__link">
    <span class="md-ellipsis">
      Run a single-target analysis
    </span>
  </a>
  
</li>
      
        <li class="md-nav__item">
  <a href="#run-a-multi-target-analysis" class="md-nav__link">
    <span class="md-ellipsis">
      Run a multi-target analysis
    </span>
  </a>
  
</li>
      
    </ul>
  
</nav>
                  </div>
                </div>
              </div>
            
          
          
            <div class="md-content" data-md-component="content">
              <article class="md-content__inner md-typeset">
                
                  



<h1 id="starting-an-analysis">Starting an analysis<a class="headerlink" href="#starting-an-analysis" title="Permanent link">&para;</a></h1>
<p>Once ViroConstrictor is <a href="../installation/">installed</a> you can use it to analyse your NGS FastQ data.  </p>
<p>You can run an analysis in two ways:</p>
<ol>
<li>A <a href="#run-a-single-target-analysis">single-target</a> analysis with run-wide parameters (the information that you give is applied to every sample in the analysis)</li>
<li>A <a href="#run-a-multi-target-analysis">multi-target</a> analysis where you provide a samplesheet (in excel, .csv or .tsv format) which contains information to run the analysis with different settings for every sample.</li>
</ol>
<p>Please see <code>viroconstrictor -h</code> for all command-line options</p>
<hr />
<h2 id="overview-of-command-line-options">Overview of command-line options<a class="headerlink" href="#overview-of-command-line-options" title="Permanent link">&para;</a></h2>
<p>Below, you can find a brief summary if all available command line options and their meaning.</p>
<table>
<thead>
<tr>
<th>Command</th>
<th>Argument</th>
<th>Explanation</th>
</tr>
</thead>
<tbody>
<tr>
<td><code>--input</code> /<br><code>-i</code></td>
<td>[input directory]</td>
<td>This is the folder containing your input fastq files.<br>Both absolute as relative paths are supported.</td>
</tr>
<tr>
<td><code>--output</code> /<br><code>-o</code></td>
<td>[output directory]</td>
<td>This is the desired output folder with all data and results.<br>If the desired folder doesn't exist yet then it will be created.</td>
</tr>
<tr>
<td><code>--reference</code> /<br><code>-ref</code></td>
<td>[reference fasta]</td>
<td>Input reference sequence genome in FASTA format</td>
</tr>
<tr>
<td><code>--primers</code> /<br><code>-pr</code></td>
<td>[primer fasta] / <code>NONE</code></td>
<td>Used primer sequences in FASTA format.<br>Use <code>NONE</code> if your sequencing protocol does not use primers</td>
</tr>
<tr>
<td><code>--features</code> /<br><code>-gff</code></td>
<td>[GFF file] / <code>NONE</code></td>
<td>The GFF3 file containing the genomic features (open reading frames etc.) matching the given reference fasta.<br>Use <code>NONE</code> if you don't have access to a GFF3 file containing this information.</td>
</tr>
<tr>
<td><code>--platform</code></td>
<td><code>nanopore</code> / <code>illumina</code> / <code>iontorrent</code></td>
<td>The sequencing platform that was used to generate the dataset. Either being 'nanopore', 'illumina' or 'iontorrent'.<br>Default is <code>nanopore</code></td>
</tr>
<tr>
<td><code>--amplicon-type</code> /<br><code>-at</code></td>
<td><code>end-to-end</code> / <code>end-to-mid</code> / <code>fragmented</code></td>
<td>The amplicon type that matches your sequencing experiment/protocol.<br>Either being <code>end-to-end</code>,<code>end-to-mid</code>, or <code>fragmented</code></td>
</tr>
<tr>
<td><code>--min-coverage</code> /<br><code>-mc</code></td>
<td>Minimum coverage</td>
<td>The minimum coverage for the consensus sequence(s)<br>Default is <strong>30</strong></td>
</tr>
<tr>
<td><code>--primer-mismatch-rate</code>/<br><code>-pmr</code></td>
<td>Fraction of maximum primer mismatches</td>
<td>The maximum percentage mismatches that is tolerated during the primer search. Mismatches are counted in substitutions between primer and reference. Insertions and/or deletions are not taken into account.<br>Default is <strong>0.1</strong> (10%)<br>This means that max 10% of the length of a primer may be a mismatch, i.e. if your primer is 30nt then maximum 3 mismatches are allowed</td>
</tr>
<tr>
<td><code>--target</code> /<br><code>--preset</code></td>
<td>Name of the viral target</td>
<td>The basic descriptive name of the viral target that is being analysed<br>i.e. "Measles", "Sars-CoV-2", "HPV16", or "Influenza_A"<br>This viral target will be used as an analysis-preset if there is a preset available for the given viral target. If no preset is available for the given target then default settings will be used. Please also see the <a href="../presets/">information regarding presets</a>.<br>Disable the use of analysis-presets with the <code>--disable-presets</code> flag.</td>
</tr>
<tr>
<td><code>--disable-presets</code> /<br><code>-dp</code></td>
<td>N/A</td>
<td>Switches off the use of analysis presets, default analysis settings will be used for all given samples/viral-targets. It is still necessary to provide the information of viral-target.</td>
</tr>
<tr>
<td><code>--match-ref</code> /<br><code>-mr</code></td>
<td>N/A</td>
<td>Enables the match-ref process for all samples in the analysis. Please see the additional information regarding the <a href="../multi-reference/#choose-the-best-reference-for-each-sample">match reference process</a></td>
</tr>
<tr>
<td><code>--segmented</code> /<br><code>-seg</code></td>
<td>N/A</td>
<td>Indicates that samples to be analyzed are segmented instead of a single reference genome. <br> Setting is only applicable to the match-ref process. Please see the addtional information regarding the <a href="../multi-reference/#choose-the-best-reference-for-each-sample-with-segmented-viruses">match reference process for segmented viruses</a></td>
</tr>
<tr>
<td><code>--threads</code> /<br><code>-t</code></td>
<td>Amount of threads</td>
<td>Number of local threads that are available to use.<br>Default is the number of available threads in your system</td>
</tr>
<tr>
<td><code>--dryrun</code></td>
<td>N/A</td>
<td>Run the ViroConstrictor workflow without actually doing anything.<br>(default: False)</td>
</tr>
<tr>
<td><code>--skip-updates</code></td>
<td>N/A</td>
<td>Skip the check for a new version</td>
</tr>
<tr>
<td><code>--version</code> /<br><code>-v</code></td>
<td>N/A</td>
<td>Shows the current version of ViroConstrictor and exits</td>
</tr>
<tr>
<td><code>--help</code> /<br><code>-h</code></td>
<td>N/A</td>
<td>Show the ViroConstrictor help document and exits</td>
</tr>
</tbody>
</table>
<h2 id="run-a-single-target-analysis">Run a single-target analysis<a class="headerlink" href="#run-a-single-target-analysis" title="Permanent link">&para;</a></h2>
<p>When running ViroConstrictor as a single-target analysis its possible to provide all necessary information via the command line. The given information will be applied to all samples in the analysis.</p>
<p>To run an analysis, you need to provide (at least) the following inputs/information:  </p>
<ul>
<li>The input directory containing your FastQ data via the <code>--input</code>/<code>-i</code> flag</li>
<li>The desired output directory for results and data via the <code>--output</code>/<code>-o</code> flag</li>
<li>The path to a reference fasta file via the <code>--reference</code>/<code>-ref</code> flag</li>
<li>
<p>Used sequencing primers <a href="../input-formatting/#formatting-your-input-primers">in correct fasta format</a> or as a bed-file via the <code>--primers</code>/<code>-pr</code> flag  </p>
<div class="admonition info">
<p>If you did not use any primers during your sequencing run then provide <code>--primers NONE</code> on the command line</p>
</div>
</li>
<li>
<p>A GFF file containing the genomic features of the given reference file via the <code>--features</code>/<code>-gff</code> flag</p>
<div class="admonition info">
<p>If you don't have access to a GFF file with genomic features matching the reference then provide <code>--features NONE</code>. ViroConstrictor will then attempt to guess the genomic features of the given reference during analysis.  </p>
<p>However, amino acid translations of the genomic features will <strong>not</strong> be provided in the results folder when no GFF file is given, if amino acid translations are desired then an input GFF file matching your reference fasta is required. </p>
</div>
</li>
<li>
<p>The used sequencing platform via <code>--platform</code>, can be "nanopore", "illumina", or "iontorrent"</p>
</li>
<li>
<p>The <a href="../amplicons/">Amplicon-type</a> in your provided data via the <code>--amplicon-type</code>/<code>-at</code> flag.  </p>
<div class="admonition info">
<p>If you did not use any primers, and set <code>--primers NONE</code> then this information will be ignored <br />
(you still have to provide this command-line flag but it won't be used)</p>
</div>
</li>
<li>
<p>A named viral target via the <code>--target</code> or <code>--preset</code> flag.</p>
<div class="admonition attention">
<p>The given viral target will be used to enable an analysis-preset if ViroConstrictor is able to match your input to a known preset. If no preset is available then the default analysis settings will be used for the given viral-target. Please also see <a href="../presets/">the documentation about working with presets</a>.<br />
You can disable the use of analysis-presets with the <code>--disable-presets</code> flag.</p>
</div>
</li>
</ul>
<p>Additional information that you can provide, but is not always required is the minimum coverage level (<code>--min-coverage</code>) as well as the primer-mismatch rate (<code>--primer-mismatch-rate</code>). When these flags are not provided then their default values are used during analysis.</p>
<div class="admonition example">
<p class="admonition-title">You can start a single-target analysis with a command such as the following:</p>
<div class="highlight"><pre><span></span><code>viroconstrictor<span class="w"> </span><span class="se">\</span>
<span class="w">    </span>--input<span class="w"> </span><span class="o">{</span>path/to/FastQ-files<span class="o">}</span><span class="w"> </span><span class="se">\</span>
<span class="w">    </span>--output<span class="w"> </span><span class="o">{</span>path/to/desired/output/folder<span class="o">}</span><span class="w"> </span><span class="se">\</span>
<span class="w">    </span>--reference<span class="w"> </span><span class="o">{</span>path/to/reference.fasta<span class="w"> </span>/<span class="w"> </span>NONE<span class="o">}</span><span class="w"> </span><span class="se">\</span>
<span class="w">    </span>--primers<span class="w"> </span><span class="o">{</span>path/to/primers.fasta<span class="w"> </span>/<span class="w"> </span>NONE<span class="o">}</span><span class="w"> </span><span class="se">\</span>
<span class="w">    </span>--features<span class="w"> </span><span class="o">{</span>path/to/features.gff<span class="w"> </span>/<span class="w"> </span>NONE<span class="o">}</span><span class="w"> </span><span class="se">\</span>
<span class="w">    </span>--platform<span class="w"> </span><span class="o">{</span>nanopore/illumina/iontorrent<span class="o">}</span><span class="w"> </span><span class="se">\</span>
<span class="w">    </span>--amplicon-type<span class="w"> </span><span class="o">{</span>end-to-end/end-to-mid<span class="o">}</span><span class="w"> </span><span class="se">\</span>
<span class="w">    </span>--min-coverage<span class="w"> </span><span class="o">{</span>coverage<span class="w"> </span>level<span class="o">}</span><span class="w"> </span><span class="se">\</span>
<span class="w">    </span>--primer-mismatch-rate<span class="w"> </span><span class="o">{</span>mismatch<span class="w"> </span>rate<span class="o">}</span><span class="w"> </span><span class="se">\</span>
<span class="w">    </span>--target<span class="w"> </span>viral-target
</code></pre></div>
</div>
<h2 id="run-a-multi-target-analysis">Run a multi-target analysis<a class="headerlink" href="#run-a-multi-target-analysis" title="Permanent link">&para;</a></h2>
<p>When running ViroConstrictor for multiple targets in a single analysis. Or if you want to set different analysis settings for every sample then you can run ViroConstrictor in multi-target analysis mode.<br />
A single analysis with different settings for each sample can be started by providing a samplesheet in Excel, CSV or TSV format.</p>
<p>An example table can be seen below, or download the example excel spreadsheet <a href="../samples.xlsx">here</a></p>
<table>
<thead>
<tr>
<th>Sample</th>
<th>Virus</th>
<th>Match-ref</th>
<th>Segmented</th>
<th>Primers</th>
<th>Reference</th>
<th>Features</th>
<th>min-coverage</th>
<th>primer-mismatch-rate</th>
</tr>
</thead>
<tbody>
<tr>
<td>Sample_1</td>
<td>SARS-CoV-2</td>
<td>FALSE</td>
<td>FALSE</td>
<td>/path/to/sars_cov_2_primers.fasta</td>
<td>/path/to/sars_cov_2_reference.fasta</td>
<td>/path/to/sars_cov_2.gff</td>
<td>50</td>
<td>0.1</td>
</tr>
<tr>
<td>Sample_2</td>
<td>Measles</td>
<td>FALSE</td>
<td>FALSE</td>
<td>/path/to/Measles_primers.fasta</td>
<td>/path/to/measles_reference.fasta</td>
<td>/path/to/measles.gff</td>
<td>30</td>
<td>0.15</td>
</tr>
<tr>
<td>Sample_3</td>
<td>Influenza_A</td>
<td>TRUE</td>
<td>TRUE</td>
<td>/path/to/Influenza-A_primers.fasta</td>
<td>/path/to/Influenza-A_reference.fasta</td>
<td>NONE</td>
<td>30</td>
<td>0.1</td>
</tr>
<tr>
<td>Sample_4</td>
<td>HPV</td>
<td>TRUE</td>
<td>FALSE</td>
<td>NONE</td>
<td>/path/to/HPV_reference.fasta</td>
<td>NONE</td>
<td>10</td>
<td>0.1</td>
</tr>
</tbody>
</table>
<p>Please keep in mind that the Sample key as given in de samplesheet must correspond with the FastQ filename in your input directory. If this isn't the case then ViroConstrictor will let you know without running to ensure you won't get unexplainable data.</p>
<p>The Virus column in the samplesheet will be used as the viral-target name for analysis. This named viral target will be used to enable an analysis-preset if ViroConstrictor is able to match your input to a known preset for analysis. If no preset is available then the default analysis settings will be used for the given viral-target.<br />
Please also see <a href="../presets/">the documentation about working with presets</a>.<br />
You can disable the use of analysis-presets with the <code>--disable-presets</code> flag.</p>
<div class="admonition example">
<p class="admonition-title">You can start a multi-target analysis with a command such as the following</p>
<div class="highlight"><pre><span></span><code>viroconstrictor<span class="w"> </span><span class="se">\</span>
<span class="w">    </span>--input<span class="w"> </span><span class="o">{</span>path/to/FastQ-files<span class="o">}</span><span class="w"> </span><span class="se">\</span>
<span class="w">    </span>--output<span class="w"> </span><span class="o">{</span>path/to/desired/output/folder<span class="o">}</span><span class="w"> </span><span class="se">\</span>
<span class="w">    </span>--samplesheet<span class="w"> </span><span class="o">{</span>path/to/samplesheet.xlsx<span class="o">}</span><span class="w"> </span><span class="se">\</span>
<span class="w">    </span>--platform<span class="w"> </span><span class="o">{</span>nanopore/illumina/iontorrent<span class="o">}</span><span class="w"> </span><span class="se">\</span>
<span class="w">    </span>--amplicon-type<span class="w"> </span><span class="o">{</span>end-to-end/end-to-mid<span class="o">}</span>
</code></pre></div>
</div>
<p>With a multi-target analysis it's also possible to combine run-wide parameters with a samplesheet.<br />
For example, you wish to analyse a batch of samples and set different minimum coverage values for every sample, but all other parameters in your analysis are the same.<br />
You can then leave out certain information in your samplesheet and supplement this information through the command line.<br />
Please note that in the given samplesheet, the columns "Samples", "Virus", and "Reference" are considered to be mandatory. starting an analysis without this information in the samplesheet will not work.</p>












                
              </article>
            </div>
          
          
<script>var target=document.getElementById(location.hash.slice(1));target&&target.name&&(target.checked=target.name.startsWith("__tabbed_"))</script>
        </div>
        
          <button type="button" class="md-top md-icon" data-md-component="top" hidden>
  
  <svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24"><path d="M13 20h-2V8l-5.5 5.5-1.42-1.42L12 4.16l7.92 7.92-1.42 1.42L13 8z"/></svg>
  Back to top
</button>
        
      </main>
      
        <footer class="md-footer">
  
  <div class="md-footer-meta md-typeset">
    <div class="md-footer-meta__inner md-grid">
      <div class="md-copyright">
  
    <div class="md-copyright__highlight">
      Copyright © 2021 RIVM
    </div>
  
  
    Made with
    <a href="https://squidfunk.github.io/mkdocs-material/" target="_blank" rel="noopener">
      Material for MkDocs
    </a>
  
</div>
      
    </div>
  </div>
</footer>
      
    </div>
    <div class="md-dialog" data-md-component="dialog">
      <div class="md-dialog__inner md-typeset"></div>
    </div>
    
    
    <script id="__config" type="application/json">{"base": "..", "features": ["navigation.tabs", "navigation.tabs.sticky", "navigation.instant", "navigation.top", "content.code.copy"], "search": "../assets/javascripts/workers/search.f8cc74c7.min.js", "translations": {"clipboard.copied": "Copied to clipboard", "clipboard.copy": "Copy to clipboard", "search.result.more.one": "1 more on this page", "search.result.more.other": "# more on this page", "search.result.none": "No matching documents", "search.result.one": "1 matching document", "search.result.other": "# matching documents", "search.result.placeholder": "Type to start searching", "search.result.term.missing": "Missing", "select.version": "Select version"}, "version": {"provider": "mike"}}</script>
    
    
      <script src="../assets/javascripts/bundle.c8b220af.min.js"></script>
      
    
  </body>
</html>